crypto/fipsmodule/ec/p256.c

```diff
@@ -35,6 +35,10 @@
 #pragma warning(disable: 4244)
 // 'initializing': conversion from 'size_t' to 'fiat_p256_limb_t'
 #pragma warning(disable: 4267)
+// 'function': incompatible types - from 'uint32_t *' to 'Limb *'
+// 'function': incompatible types - from 'fiat_p256_felem' to 'const Limb *'
+// etc.
+#pragma warning(disable: 4133)
```

Comment:
This is probably a bad idea, as I'm not sure if it's okay to use 32-bit math on ARM64 in this module

Reply:
Yes, we shouldn't disable these warnings. When the code is correct, then there will be no warnings. These warnings help us ensure we're not accidentally using 32-bit integer types instead of 32-bit integer types or vice versa. If there are cases we need to do conversions then we should add explicit casts for those specific instances.

Reply:
Okay, it looks like we have a problem then because of the `__uint128_t` type 😅

Reply:
@briansmith  What do you think about including a bigint library? For example boost's [cpp_int](https://www.boost.org/doc/libs/1_76_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints/cpp_int.html) or [calccrypto](https://github.com/calccrypto/uint128_t)? That could solve all the issues with math at once, and we could only include it on Windows ARM64.

Reply:
> Okay, it looks like we have a problem then because we don't have the __uint128_t type, and we also don't have a OPENSSL_USE_NISTZ256 implementation.

If I understand what you're saying correctly, there is another PR that adds the fallback `bn_mul_mont` implementation that would be needed for supporting MSVC to work.

HOWEVER, that code will be much slower than using the uint128_t-based arithmetic from Fiat Crypto that using clang would provide.

So, with that in mind, maybe we should first get this working with `TARGET_CC=clang` and then worry about supporting MSVC as the C compiler in a separate PR?


> @briansmith What do you think about including a bigint library? For example boost's cpp_int or calccrypto? That could solve all the issues with math at once, and we could only include it on Windows ARM64.

It would be very unlikely that we'd add an external dependency for the math.

Reply:
One of the main goals for me is to make sure Rustup can compile with rustls-tls feature enabled. If we can somehow specify in build.rs that we want to use clang.exe as the C compiler, and that setting can be dragged into Rustup, I'm generally happy 🙂

Reply:
@briansmith Okay, before I change a ton of `#if defined(__GNUC__)` to `#if defined(__GNUC__) || defined(__clang__)` in this context:

```c
#if defined(__GNUC__) || defined(__clang__)
#pragma GCC diagnostic ignored "-Wsign-conversion"
#pragma GCC diagnostic ignored "-Wconversion"
#endif
```

-- is this an expected change? 🙃

Reply:
There is some code that does that in *ring*, because BoringSSL does that. But only code that we share with BoringSSL is supposed to do that, ideally.

For code not shared with BoringSSL, we should be using explicit casts and avoid disabling the warnings.

If you're going to hand this off to me, then go ahead and disable whatever warnings and then I'll fix that.

Reply:
I tried to be as conservative as I could when introducing those `#if ... defined(__clang__)` 🙂 It now builds on my SPX and all the tests run. I'm going to try cross-compiling everything from an x64 machine now, and if that succeeds, I think my job here is largely done 🙂 I'd prefer to leave the code itself to you since it likely has nothing to do with Windows ARM64 target but rather with the logic itself, and I don't trust myself there as I have zero knowledge of crypto 😬